File: 1/main.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"index/suffixarray"
	"os"
	"sort"
	"strconv"
	"strings"

	"github.com/scbizu/aoc2022/helper/input"
)

func main() {
	phase1()
	phase2()
}

func phase1() {
	txt := input.NewTXTFile("./input.txt")
	ctx := context.Background()
	var total int
	txt.ReadByLineEx(ctx, func(_ int, line string) error {
		if line == "" {
			return nil
		}
		number := bytes.NewBuffer(nil)
		for _, c := range line {
			if '0' < c && c <= '9' {
				number.WriteByte(byte(c))
			}
		}
		numberStr := bytes.NewBuffer(nil)
		for i := 0; i < number.Len(); i++ {
			if i == 0 {
				numberStr.WriteByte(number.Bytes()[i])
				continue
			}
			if i == number.Len()-1 {
				numberStr.WriteByte(number.Bytes()[i])
				continue
			}
		}
		if numberStr.Len() == 1 {
			numberStr.WriteByte(numberStr.Bytes()[0])
		}
		n, _ := strconv.Atoi(numberStr.String())
		total += n
		return nil
	})
	fmt.Fprintf(os.Stdout, "phase 1: %d\n", total)
}

var digitMap = map[string]byte{
	"one":   '1',
	"two":   '2',
	"three": '3',
	"four":  '4',
	"five":  '5',
	"six":   '6',
	"seven": '7',
	"eight": '8',
	"nine":  '9',
}

type digit struct {
	index int
	len   int
	value byte
}

func phase2() {
	txt := input.NewTXTFile("./input.txt")
	ctx := context.Background()
	var total int
	txt.ReadByLineEx(ctx, func(_ int, line string) error {
		if line == "" {
			return nil
		}
		var indexes []digit
		for k, v := range digitMap {
			if !strings.Contains(line, k) {
				continue
			}
			index := suffixarray.New([]byte(line))
			res := index.Lookup([]byte(k), -1)
			for _, r := range res {
				indexes = append(indexes, digit{
					index: r,
					len:   len(k),
					value: v,
				})
			}
		}
		for index, c := range line {
			if '0' < c && c <= '9' {
				indexes = append(indexes, digit{
					index: index,
					len:   1,
					value: byte(c),
				})
			}
		}
		sort.Slice(indexes, func(i, j int) bool {
			return indexes[i].index < indexes[j].index
		})
		number := bytes.NewBuffer(nil)
		for i := 0; i < len(indexes); i++ {
			number.WriteByte(indexes[i].value)
		}
		numberStr := bytes.NewBuffer(nil)
		for i := 0; i < number.Len(); i++ {
			if i == 0 {
				numberStr.WriteByte(number.Bytes()[i])
				continue
			}
			if i == number.Len()-1 {
				numberStr.WriteByte(number.Bytes()[i])
				continue
			}
		}
		if numberStr.Len() == 1 {
			numberStr.WriteByte(numberStr.Bytes()[0])
		}
		n, err := strconv.Atoi(numberStr.String())
		if err != nil {
			panic(err)
		}
		fmt.Printf("numberStr: %s, n: %d\n", number.String(), n)
		total += n
		return nil
	})
	fmt.Fprintf(os.Stdout, "phase 2: %d\n", total)
}

File: 1/main.zig
const std = @import("std");

const IndexNumber = struct {
    index: u32,
    number: u8,
};

const Number = struct {
    numberStr: []u8,
    number: u8,
};

fn indexNumberAscending(lhs: IndexNumber, rhs: IndexNumber) bool {
    return lhs.index < rhs.index;
}

pub fn main() !void {
    var file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();
    var reader = std.io.bufferedReader(file.reader());
    var stream = reader.reader();
    var buffer: [1024]u8 = undefined;
    var total: u32 = 0;
    const numberSet = std.ArrayList(Number).init(std.heap.page_allocator);
    defer numberSet.deinit();
    numberSet.insert(.{ .numberStr = "one", .number = 1 });
    numberSet.insert(.{ .numberStr = "two", .number = 2 });
    numberSet.insert(.{ .numberStr = "three", .number = 3 });
    numberSet.insert(.{ .numberStr = "four", .number = 4 });
    numberSet.insert(.{ .numberStr = "five", .number = 5 });
    numberSet.insert(.{ .numberStr = "six", .number = 6 });
    numberSet.insert(.{ .numberStr = "seven", .number = 7 });
    numberSet.insert(.{ .numberStr = "eight", .number = 8 });
    numberSet.insert(.{ .numberStr = "nine", .number = 9 });
    while (try stream.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
        var array = std.ArrayList(IndexNumber).init(std.heap.page_allocator);
        defer array.deinit();
        for (line, 0..) |c, i| {
            if (c >= '0' and c <= '9') {
                try array.append(.{ .index = i, .number = c - '0' });
            }
        }
        for (numberSet.items) |number| {
            // if line contains number
            const index = try std.mem.indexOf([]u8, line, number.numberStr);
            if (index != null) {
                try array.append(.{ .index = index, .number = number.number });
            }
        }

        if (array.items.len == 1) {
            try array.append(array.items[0]);
        }
        std.sort.insertion(IndexNumber, array.items, {}, indexNumberAscending);
        const str = [_]u8{ array.items[0], array.items[1] };
        const num = try std.fmt.parseInt(u32, &str, 10);
        total += num;
    }
    std.debug.print("total: {d}\n", .{total});
}
